x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/TransportVersionClusterStateUpgradeIT.java

```diff
@@ -12,70 +12,135 @@
+            case OLD -> {
+                if (UPGRADE_FROM_VERSION.before(VERSION_INTRODUCING_TRANSPORT_VERSIONS)) {
+                    assertFalse(description, hasTransportVersions);
+                    assertFalse(description, hasNodesVersions);
+                } else if (UPGRADE_FROM_VERSION.before(VERSION_INTRODUCING_NODES_VERSIONS)) {
+                    assertTrue(description, hasTransportVersions);
+                    assertFalse(description, hasNodesVersions);
+                } else {
+                    assertFalse(description, hasTransportVersions);
+                    assertTrue(description, hasNodesVersions);
+                }
+            }
+            case MIXED -> {
+                if (UPGRADE_FROM_VERSION.before(VERSION_INTRODUCING_TRANSPORT_VERSIONS)) {
+                    assertFalse(description, hasTransportVersions);
+                } else if (UPGRADE_FROM_VERSION.before(VERSION_INTRODUCING_NODES_VERSIONS)) {
+                    assertTrue(description, hasNodesVersions || hasTransportVersions);
+                } else {
+                    assertFalse(description, hasTransportVersions);
+                    assertTrue(description, hasNodesVersions);
+                }
+            }
+            case UPGRADED -> {
+                assertFalse(description, hasTransportVersions);
+                assertTrue(description, hasNodesVersions);
+                assertTrue(description, versionsByNodeId.values().stream().allMatch(v -> v.equals(Version.CURRENT)));
             }
         }
-    }
 
-    public void testCompletesRealTransportVersions() throws IOException {
-        assumeTrue("TransportVersion introduced in 8.8.0", UPGRADE_FROM_VERSION.before(Version.V_8_8_0));
-        assumeTrue(
-            "This only has visible effects when upgrading beyond 8.8.0",
-            TransportVersion.current().after(TransportVersions.V_8_8_0)
-        );
-        assumeTrue("Only runs on the upgraded cluster", CLUSTER_TYPE == ClusterType.UPGRADED);
-        // once everything is upgraded, the master should fill in the real transport versions
+        if (hasTransportVersions) {
+            assertTrue(description, UPGRADE_FROM_VERSION.before(VERSION_INTRODUCING_NODES_VERSIONS));
+            assertTrue(description, UPGRADE_FROM_VERSION.onOrAfter(VERSION_INTRODUCING_TRANSPORT_VERSIONS));
+            assertNotEquals(description, ClusterType.UPGRADED, CLUSTER_TYPE);
 
-        Request request = new Request("GET", "/_cluster/state/nodes");
-        Map<String, Object> response = entityAsMap(client().performRequest(request));
-        Map<?, TransportVersion> tvs = ((List<?>) response.get("transport_versions")).stream()
-            .map(o -> (Map<?, ?>) o)
-            .collect(Collectors.toMap(m -> m.get("node_id"), m -> TransportVersion.fromString(m.get("transport_version").toString())));
+            assertEquals(description, nodeIds.size(), clusterState.evaluateArraySize("transport_versions"));
```

Comment:
Can use `hasSize` matcher here

Reply:
Not without doing some unchecked casting or reversing the expected/actual values I think.